### ISSUE TITLE
Coverage report for --cov argument

### ISSUE THREAD

**Issue by tomdye** [dojo/cli-test-intern#15]
**Bug** 

`--cov` option does not seem to be working.

**Comment by agubler** [comment]
Coverage seems to be only working for functionals and reports to the console regardless of the whether the `--cov` option is passed.

However they output is not remapped back to source (obviously), which we really need to be included for `dojo test`, currently just get the report for the `main.js` webpack bundle.

Lots of things going on here that need to be addressed, @kitsonk do you think this is something you could look at? Specifically the remapping part (as I know you've had some experience there 😄 ).

So in summary

1) Unit tests need to generate coverage (as well as functionals and combine if they are both run)
2) The coverage needs to be remapped back to source.
3) Potentially fail if a coverage threshold is not met.

**Comment by kitsonk** [comment]
Sure, I can take a look.

**Comment by kitsonk** [comment]
> 3) Potentially fail if a coverage threshold is not met.

Where or how would we express that?

**Comment by agubler** [comment]
I was pretty much thinking off the top of my head so haven't really thought it through too much! Are you asking where we'd express the threshold configuration?

**Comment by matt-gadd** [comment]
@kitsonk @agubler seems perfect for the `.dojorc`

**Comment by agubler** [comment]
That's what I was thinking @matt-gadd

**Comment by kitsonk** [comment]
Ok, I think I got my head around this.  A few thoughts though:

- We will need to implement a custom reporter, like we have in `grunt-dojo2`.  While Intern has a JSON coverage reporter now, all the other reporters will automatically output unmapped coverage console blocks.  Which will be confusing.  This then does open up if we are happy with the output of existing reporter.  Which currently outputs something like this:
    ```
    ⠸ Running tests✓ chrome on any platform - TodoMVC - Dojo 2 - Counter - should display the 
    current number of todo items (0.481s)
    ⠴ Running tests✓ chrome on any platform - TodoMVC - Dojo 2 - Clear completed button - should 
    display the correct text (0.18s)
    ```
    Which ain't very pretty.  The one with `grunt-dojo2` is likely a bit terse for `dojo test`.  Any thoughts?
- `dojo test -c local` which is allegedly is the default fails.  I will open another issue for that.
- Right now `--cov` is a bit confusing in that you get the coverage table irrespective of if you use it. Basically we want it to output a mapped coverage HTML report?  Also, I would suggest that we output a `coverage.json` and a `coverage.lcov` which are what are used by things like codecov.io.  Also, I assume we still want a covered remapped table output by default, or do we really want there to be no instrumentation of the project with `dojo test` and only go through that process with `dojo test --cov`.

**Comment by matt-gadd** [comment]
> We will need to implement a custom reporter, like we have in grunt-dojo2. While Intern has a JSON coverage reporter now, all the other reporters will automatically output unmapped coverage console blocks. Which will be confusing. This then does open up if we are happy with the output of existing reporter. 

It does seem like we might want to look into creating a reporter with a nicer output for `dojo test`, what do you think?

> Right now --cov is a bit confusing in that you get the coverage table irrespective of if you use it. Basically we want it to output a mapped coverage HTML report? Also, I would suggest that we output a coverage.json and a coverage.lcov which are what are used by things like codecov.io. Also, I assume we still want a covered remapped table output by default, or do we really want there to be no instrumentation of the project with dojo test and only go through that process with dojo test --cov.

It seems to me like ideally we would want to support coverage by default out of the box? I can't think why this would be an optional argument in the first place (but to be honest, some of the default options and arguments themselves have been unusual so far in `dojo test`, see https://github.com/dojo/cli-test-intern/pull/12)

**Comment by kitsonk** [comment]
> I can't think why this would be an optional argument in the first place

I think writing out the physical reports to disk wouldn't necessarily be a default?  Or at least we should maybe put them someplace other than the root of the project?  With `grunt-dojo2` we do output them but we have our ignores in place to make sure they don't get checked in.  I don't know if `dojo create` stubs out ignores, but we should either put them in a path that is always ignored, or ensure they are ignored by default.  Also we might want to do a better job of informing the user that they have a nice handy report for them.  I don't know if there is an easy logical way to tell them how to access it to.  Since we bundle a server, we might want to consider serving up the coverage report.

**Comment by matt-gadd** [comment]
As part of `cli-build` we output to a directory (`dist`), so I don't see why this would potentially be any different. I don't think create-app currently presumes `git`, so technically their is nothing to ignore. Either way, it seems like if this is an issue we should make it a `cli-create-app` one?

As far as i'm aware the html-report is static and can be loaded from the file system fine, so we can potentially just log to the console a link to that from the task?

**Comment by kitsonk** [comment]
Another thought I had:

- We need to clear up some terminology too and what we expect.  I think we have conflated "unit" tests and "functional" tests with the environments they run in.  With `grunt-dojo2` this is what occurs:
  - `grunt test` - Run the tests under node, functional tests do not run in that environment, so only unit tests run.
  - `grunt test:local` - Run the tests against the local tunnel, both unit tests and functional tests run because the environment supports it.
  - `grunt test:saucelabs` - Like local but against all the environments specified in the configuration.
  - When we test under CI, we essentially run two sets of tests, we run them against node and we run them against SauceLabs or BrowserStack.  We collect both sets of coverage information, merge them and remap them into a single set of coverage information.

What is sort of confusing is that `dojo test` seems to try to run the test suites under node, of course which only the unit ones.  `dojo test -f` seems to be like `grunt test:local` which is says it only tests the functional suites, but I suspect it does unit and functional through the local tunnel.  `dojo test -a` seems to be a combination of running against node and running against the local tunnel.

So do we a) really only run unit or functional or b) really trying to choose the environment we are testing in?  I think it is really the latter, where by default it is just under node, we can choose local tunnel or we can choose a combined set, where we do it in node and local tunnel.

Another whole thought, do we want to make it even easier and install chromedriver for the user and bootstrap it when we are running tests against the local tunnel?  There are challenges in that we wouldn't be able to easily control what sort of version of chrome they have installed locally and there could be a mismatch between the driver and the version of chrome.  There maybe also other environmental considerations which might not make it "automagic"

**Comment by agubler** [comment]
To be honest I have also always found the options for `dojo test` a little confusing, but it is slightly better in that it now only runs the units (in node) by default.

I'm happy if we raise an issue to further clear up the options and testing behaviour.

As for automatically installing chromedriver, do we need that now intern supports the seleniumTunnel and automatically downloads it when running functimnals locally?

**Comment by kitsonk** [comment]
> As for automatically installing chromedriver, do we need that now intern supports the seleniumTunnel and automatically downloads it when running functionals locally?

DOH! I had been doing chromedriver for so long I never realised it was already automagic. 😭

**Comment by kitsonk** [comment]
Ok, so what I am going to do for this issue and subsequent PR:

- Implement a custom reporter
- Remap the coverage from whatever tests are run
- Output the coverage table to console by default
- Generate the `output/test/html-report`, `output/test/coverage.json`, and `output/test/coverage.lcov` when the `--cov` flag is passed

I won't be addressing:

- Changing the behaviour of what gets tested with the `-f`, `-u` and `-a` flags.
- Integrating `@dojo/intern-helper`
- Fixing issue #18 (`dojo test -c local` failing)
